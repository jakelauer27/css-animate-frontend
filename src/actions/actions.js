export const signIn = (user) => ({
  type: 'SIGN_IN',
  user
})

export const signOut = () => ({
  type: 'SIGN_OUT'
})

export const loadPrebuiltAnimations = (prebuiltAnimations) => ({
  type: 'LOAD_PREBUILT_ANIMATIONS',
  prebuiltAnimations
})

export const updateCurrentAnimation = (animation) => ({
  type: 'UPDATE_CURRENT_ANIMATION',
  animation
})

export const saveOriginalAnimation = (animation) => ({
  type: 'SAVE_ORIGINAL_ANIMATION',
  animation
})

export const loadMyAnimations = (myAnimations) => ({
  type: 'LOAD_MY_ANIMATIONS',
  myAnimations
})

export const deleteMyAnimation = (id) => ({
  type: 'DELETE_MY_ANIMATION',
  id
})

export const selectAnimationForEdit = (animation) => ({
  type: 'SELECT_ANIMATION_FOR_EDIT',
  animation
})

export const removeAnimationForEdit = () => ({
  type: 'REMOVE_ANIMATION_FOR_EDIT'
})


