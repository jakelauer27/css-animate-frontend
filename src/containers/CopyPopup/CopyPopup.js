import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { PropTypes } from 'prop-types'
 
export class CopyPopup extends Component {
  constructor() {
    super()
    this.state = {
      copyKeyframes: false,
      copyProperties: false,
      copyKeyframesText: 'copy',
      copyPropertiesText: 'copy'
    }
  }

  copyPasteKeyframes = () => {
    return (
      <div className='keyframes-popup'>
        <textarea className='keyframe-text' readOnly value={`@keyframes ${this.props.animation.keyframes.name} {
  ${this.getKeyframeStages(this.props.animation.keyframes)}
}`
}>
        </textarea>  
      </div>
    )
  }

  copyPasteAnimation = () => {
    let keys = Object.keys(this.props.animation.properties)
    let props = this.props.animation.properties
    let animation = (
      <div className='animation-popup'>
        <textarea className='animation-text' readOnly value={
            `animation:${
                keys.map((key, i) => {
                  if (i === keys.length -1) {
                    return ''
                  } 
                  return ` ${props[key]}`
                }).join('')}; 
animation-fill-mode: ${props['fill-mode']};`
        }>
        </textarea>
      </div>
    )
    return animation
  }

  getKeyframeStages(keyframe) {
    var obj = keyframe.sections.map( (section, i) => {
      return (
  `${section.label} {
    ${
      section.properties.map( (prop, i) => {
        if (i === 0) {
          return `${prop.name}: ${prop.value};`
        } 
        return (`
    ${prop.name}: ${prop.value};`                
          )
        }).join('')
      }
  }
  `
      )
    }) 
    return obj.join('')
  }

  copySelection = (e) => {
    e.target.previousElementSibling.childNodes[0].select()
    document.execCommand('copy')

    this.setState({
      copyKeyframes: false,
      copyProperties: false,
      copyKeyframesText: 'copy',
      copyPropertiesText: 'copy',
      [e.target.name]: true,
      [e.target.name + 'Text']: 'copied'
    })
  }

  render() {
    const { copyKeyframes, copyProperties, copyKeyframesText, copyPropertiesText } = this.state
    return (
    <div className='copy-paste-popup'>
      <div className='popup-keyframes-container popup-container'>
        <h2 className='copy-popup-label'>Keyframes</h2>
        {
          this.copyPasteKeyframes()
        } 
        <button 
          name='copyKeyframes' 
          className={`copy-code-btn copy-keyframes-btn ${copyKeyframes && 'copied'}`} 
          onClick={this.copySelection}>{copyKeyframesText}
        </button>
      </div>
      <div className='popup-animation-container popup-container'>
        <h2 className='copy-popup-label'>Animation</h2>
        {
          this.copyPasteAnimation()
        } 
        <button 
          name='copyProperties'
          className={`copy-code-btn copy-animation-btn ${copyProperties && 'copied'}`} 
          onClick={this.copySelection}>{copyPropertiesText}
        </button>
      </div>
      <Link to={`/properties`}>
        <button className='close-popup-btn'><i className="fas fa-times"></i></button>
      </Link>
    </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  animation: state.currentAnimation
})

CopyPopup.propTypes = {
  animation: PropTypes.object
}

export default connect(mapStateToProps)(CopyPopup)