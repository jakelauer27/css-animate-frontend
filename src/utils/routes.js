export const routes = [`/properties/selectAnimation`, `/properties/signup`, `/properties/login`, `/properties/howto/general`, `/keyframes`, `/properties`, '/', `/properties/howto/properties`,`/properties/howto/keyframes`, '/properties/selectAnimation/create', `/properties/copy`, '/error']