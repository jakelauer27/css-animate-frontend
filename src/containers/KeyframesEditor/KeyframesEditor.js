import React, { Component } from 'react'
import * as formValidation from '../../utils/formValidators'
import { connect } from 'react-redux'
import { updateCurrentAnimation } from '../../actions/actions'
import { updateKeyframes } from '../../utils/keyframesInsertion'
import { PropTypes } from 'prop-types'
import { uid } from 'react-uid'

export class KeyframesEditor extends Component {

  saveKeyframesStages(e) {
    formValidation.keyframeStage(e.target, e.target.value)
    document.querySelector('.stop-btn').click()
    let newAnimation = JSON.parse(JSON.stringify(this.props.animation))
    let stageLabel = e.target.classList[1]
    let stageIndex = ''
    newAnimation.keyframes.sections.forEach( (stage, i) => {
       if (stage.name === stageLabel) {
         stageIndex = i
       }
    })
    newAnimation.keyframes.sections[stageIndex].label = e.target.value || '%'
    this.props.updateCurrentAnimation(newAnimation)
    updateKeyframes(newAnimation.keyframes)
  }

  saveKeyframesProps(e) {
    formValidation.keyframeValue(e.target, e.target.value)
    document.querySelector('.stop-btn').click()
    let newAnimation = {...this.props.animation}
    let stageLabel = e.target.parentElement.parentElement.childNodes[0].classList[1]
    let propLabel = e.target.classList[1]
    let stageIndex = ''
    let propIndex = ''
    newAnimation.keyframes.sections.forEach( (stage, i) => {
       if (stage.name === stageLabel) {
         stageIndex = i
       }
    })
    newAnimation.keyframes.sections[stageIndex].properties.forEach( (prop, i) => {
       if (prop.name === propLabel) {
         propIndex = i
       }
    })
    newAnimation.keyframes.sections[stageIndex].properties[propIndex].value = e.target.value
    this.props.updateCurrentAnimation(newAnimation)
    updateKeyframes(newAnimation.keyframes)
  }

  render() {
    const { animation } = this.props
    if (!animation.keyframes) {
      return <div></div>
    }
    return (
      <div className='keyframes-container'>
        <div className='keyframe'>
          <p className='keyframes-name'><span>@keyframes </span>{animation.keyframes.name} <span> {'{'}</span></p>
          {
            animation.keyframes.sections.map( (section) => {
              return (
                <div className='keyframes-section' key={uid(section)}>
                  <input className={`keyframes-label ${section.name}`} 
                    value={section.label} 
                    type='text'
                    onChange={e => this.saveKeyframesStages(e)}/> 
                    <span> {'{'}</span>
                  {
                    section.properties.map( (prop) => {
                      return (
                        <div className='props-container' key={uid(prop)}>
                          <p className='keyframe-property'>{prop.name}<span>:</span></p>
                          <input className={`keyframe-prop-value ${prop.name}`} 
                            type='text' 
                            value={prop.value}
                            onChange={e => this.saveKeyframesProps(e)}></input>
                        </div>                   
                        )
                      })
                    }
                  <p className='close-curly'>{'}'}</p>
                </div>
              )
            }) 
          }
          <p className='close-curly'>{'}'}</p>
        </div>
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  animation: state.currentAnimation
})

export const mapDispatchToProps = (dispatch) => ({
  updateCurrentAnimation: (animation) => dispatch(updateCurrentAnimation(animation))
})

KeyframesEditor.propTypes = {
  animation: PropTypes.object.isRequired,
  updateAnimation: PropTypes.func.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(KeyframesEditor)