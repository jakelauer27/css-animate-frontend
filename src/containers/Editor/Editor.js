import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Route, Switch, NavLink, Link } from 'react-router-dom'
import KeyframesEditor from '../KeyframesEditor/KeyframesEditor'
import PropertiesEditor from '../PropertiesEditor/PropertiesEditor'
import { updateCurrentAnimation } from '../../actions/actions'
import * as CSSInsertion from '../../utils/keyframesInsertion'
import CopyPopup from '../CopyPopup/CopyPopup'
import { PropTypes } from 'prop-types'

export class Editor extends Component {

  resetAnimation(animation) {
    animation = JSON.parse(animation)
    this.props.updateCurrentAnimation(animation)
    CSSInsertion.updateKeyframes(animation.keyframes)
    this.resetInputValidation()
  }
  
  resetInputValidation() {
    document.querySelectorAll('input').forEach( input => {
      input.classList.remove('red');
    });
    document.querySelector('.play-btn').removeAttribute('disabled');
  }


  render() {
    const { currentAnimation } = this.props

    return (
      <div className='editor-component'>
        <div className='editor-container'>
          <div className='editor-options'>
            <NavLink 
              to={`/properties`}
              className='animation-btn'
              activeClassName='selected'>properties
            </NavLink>
            <NavLink 
              to={`/keyframes`}
              className='keyframes-btn'
              activeClassName='selected'>keyframes
            </NavLink>
          </div>
          <div className='editor'>
          {
          <Switch>
            <Route path={`/properties`} component={PropertiesEditor}/>
            <Route path={`/keyframes`} component={KeyframesEditor}/>
          </Switch>
          }
          </div>
        </div>
        <div className='editor-bottom-btns-container'>
          <Link to={`/properties/copy`} className='lower-btn copy-btn'>
            copy code   
          </Link>
          <Route path={`/properties/copy`} render={() => <CopyPopup currentPage={currentAnimation}/>} />
          <button className={`lower-btn reset-btn`} onClick={() => this.resetAnimation(this.props.originalAnimation)}>
            reset
          </button>
        </div>
      </div>
    )
  }
}


export const mapStateToProps = (state) => ({
  originalAnimation: state.originalAnimation,
  currentAnimation: state.currentAnimation
})

export const mapDispatchToProps = (dispatch) => ({
  updateCurrentAnimation: (animation) => dispatch(updateCurrentAnimation(animation))
})

Editor.propTypes = {
 originalAnimation: PropTypes.string,
 currentAnimation: PropTypes.object,
 updateCurrentAnimation: PropTypes.func.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(Editor)
