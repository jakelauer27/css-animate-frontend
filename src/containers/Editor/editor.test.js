import { Editor, mapStateToProps, mapDispatchToProps } from './Editor' 
import React from 'react'
import { shallow, mount } from 'enzyme'
import { loadAnimation } from '../../actions/actions'
import * as CSSInsertion from '../../utils/keyframesInsertion'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import KeyframesEditor from '../KeyframesEditor/KeyframesEditor'
import PropertiesEditor from '../PropertiesEditor/PropertiesEditor'
import CopyPopup from '../CopyPopup/CopyPopup'


jest.mock('../../utils/keyframesInsertion')

CSSInsertion.updateKeyframes.mockImplementation(() => jest.fn())

describe('Editor', () => {
  let wrapper
  let mockOriginalAnimation
  let mockUpdateCurrentAnimation
  let mockSaveOriginalAnimation
  let mockEditAnimation
  const mockLoadAnimation = jest.fn()
  const mockAnimation = { 
    'properties': {
      'name': 'slideInX',
      'duration': '1.5s',
      'timingFunction': 'ease',
      'delay': '0s',
      'iterationCount': '1',
      'direction': 'normal',
      'fillMode': 'forwards'
    },
    'keyframes': {
      'name': 'slideInX',
      'sections': [
        {
          'name': '0%',
          'label': '0%',
          'properties': [
            {
              'name': 'transform',
              'value': 'translateX(-300px)'
            }
          ]
        },
        {
          'name': '100%',
          'label': '100%',
          'properties': [
            {
              'name': 'transform',
              'value': 'translateX(0px)'
            }
          ]
        },
      ]
    }
  }

  beforeEach(() => {
    mockOriginalAnimation = '{}'
    mockEditAnimation = jest.fn()
    mockSaveOriginalAnimation = jest.fn()
    mockUpdateCurrentAnimation = jest.fn()
    wrapper = shallow(
      <Editor 
        user_id={1}
        originalAnimation={mockOriginalAnimation}
        currentAnimation={mockAnimation}
        updateCurrentAnimation={mockUpdateCurrentAnimation}
        editAnimation={mockEditAnimation}
        saveOriginalAnimation={mockSaveOriginalAnimation}
      />
    )
  })

  it('should have a default state', () => {
    const expected = {saveText: 'save'}
    expect(wrapper.state()).toEqual(expected)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should call resetAnimation when reset button is clicked', () => {
    const loadNewSpy = jest.spyOn(wrapper.instance(), 'resetAnimation')
    wrapper.find('.reset-btn').simulate('click')
    expect(loadNewSpy).toHaveBeenCalledWith(mockOriginalAnimation)
  })

  it('should call saveAnimation when save button is clicked', () => {
    const spy = jest.spyOn(wrapper.instance(), 'saveAnimation')
    wrapper.find('.save-btn').simulate('click')
    expect(spy).toHaveBeenCalled()
  })

  describe('resetInputValidation', () => {
    document.querySelectorAll = jest.fn().mockImplementation(() => {
      return [{classList: {remove: jest.fn()}}]
    })
    document.querySelector = jest.fn().mockImplementation(() => ({removeAttribute: jest.fn()}))
  
    it('should remove red class from all inputs', () => {
      wrapper.instance().resetInputValidation()
      expect(document.querySelectorAll).toHaveBeenCalled()
    })

    it('should remove disabled attr from play btn', () => {
      wrapper.instance().resetInputValidation()
      expect(document.querySelector).toHaveBeenCalled()
    })
  })

  describe('resetAnimation', () => {

  })

  describe('saveAnimation', () => {

  })

  describe('resetSaveText', () => {

  })

  describe('routes', () => {

    it('should render keyframes component if on keyframes route', () => {
      wrapper = mount(
        <Provider store={createStore(() => ({animation: mockAnimation}))}>
          <MemoryRouter initialEntries={['/slideInX/keyframes']}>
            <Editor animation={mockAnimation} loadNewAnimation={mockLoadAnimation} currentAnimation={'slideInX'}/>
          </MemoryRouter>
        </Provider>
        )
        expect(wrapper.find(KeyframesEditor)).toHaveLength(1)
    })
    
    it('should render properties component if on properties route', () => {
      wrapper = mount(
        <Provider store={createStore(() => ({animation: mockAnimation}))}>
          <MemoryRouter initialEntries={['/properties']}>
            <Editor animation={mockAnimation} loadNewAnimation={mockLoadAnimation} currentAnimation={'slideInX'}/>
          </MemoryRouter>
        </Provider>
        )
        expect(wrapper.find(PropertiesEditor)).toHaveLength(1)
    })

    it('should render copyCode component if on copyCode route', () => {
      wrapper = mount(
        <Provider store={createStore(() => ({animation: mockAnimation}))}>
          <MemoryRouter initialEntries={['/properties/copy']}>
            <Editor animation={mockAnimation} loadNewAnimation={mockLoadAnimation} currentAnimation={'slideInX'}/>
          </MemoryRouter>
        </Provider>
        )
        expect(wrapper.find(CopyPopup)).toHaveLength(1)
    })
  })

  describe('mapStateToProps', () => {

    it('should return a props object with user, oganimation, and currentanimation properties', () => {
      const mockState = {currentAnimation: mockAnimation, originalAnimation: mockOriginalAnimation, user:{id: 1}}
      const expected = {currentAnimation: mockAnimation, originalAnimation: mockOriginalAnimation, user_id: 1}
      expect(mapStateToProps(mockState)).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {

    it('should return a props object with an upDateAnimation property', () => {
        const mockDispatch = jest.fn()
        const actionToDispatch = loadAnimation(mockAnimation)
        const mappedProps = mapDispatchToProps(mockDispatch)
        mappedProps.loadNewAnimation(mockAnimation)
        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})