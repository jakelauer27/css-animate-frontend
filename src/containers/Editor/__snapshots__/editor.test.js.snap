// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Editor
    currentAnimation={
      Object {
        "keyframes": Object {
          "name": "slideInX",
          "sections": Array [
            Object {
              "label": "0%",
              "name": "0%",
              "properties": Array [
                Object {
                  "name": "transform",
                  "value": "translateX(-300px)",
                },
              ],
            },
            Object {
              "label": "100%",
              "name": "100%",
              "properties": Array [
                Object {
                  "name": "transform",
                  "value": "translateX(0px)",
                },
              ],
            },
          ],
        },
        "properties": Object {
          "delay": "0s",
          "direction": "normal",
          "duration": "1.5s",
          "fillMode": "forwards",
          "iterationCount": "1",
          "name": "slideInX",
          "timingFunction": "ease",
        },
      }
    }
    editAnimation={[MockFunction]}
    originalAnimation="{}"
    saveOriginalAnimation={[MockFunction]}
    updateCurrentAnimation={[MockFunction]}
    user_id={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="editor-container"
        >
          <div
            className="editor-options"
          >
            <NavLink
              activeClassName="selected"
              aria-current="page"
              className="animation-btn"
              to="/properties"
            >
              properties
            </NavLink>
            <NavLink
              activeClassName="selected"
              aria-current="page"
              className="keyframes-btn"
              to="/keyframes"
            >
              keyframes
            </NavLink>
          </div>
          <div
            className="editor"
          >
            <Switch>
              <Route
                component={[Function]}
                path="/properties"
              />
              <Route
                component={[Function]}
                path="/keyframes"
              />
            </Switch>
          </div>
        </div>,
        <div
          className="editor-bottom-btns-container"
        >
          <Link
            className="lower-btn copy-btn"
            replace={false}
            to="/properties/copy"
          >
            copy code
          </Link>
          <button
            className="lower-btn reset-btn"
            onClick={[Function]}
          >
            reset
          </button>
          <Link
            className="lower-btn save-as-btn"
            replace={false}
            to="/properties/saveas"
          >
            save as
          </Link>
          <button
            className="lower-btn save-btn"
            disabled={true}
            onClick={[Function]}
          >
            save
          </button>
          <Switch>
            <Route
              path="/properties/copy"
              render={[Function]}
            />
            <Route
              path="/properties/saveas"
              render={[Function]}
            />
          </Switch>
        </div>,
      ],
      "className": "editor-component",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="editor-options"
            >
              <NavLink
                activeClassName="selected"
                aria-current="page"
                className="animation-btn"
                to="/properties"
              >
                properties
              </NavLink>
              <NavLink
                activeClassName="selected"
                aria-current="page"
                className="keyframes-btn"
                to="/keyframes"
              >
                keyframes
              </NavLink>
            </div>,
            <div
              className="editor"
            >
              <Switch>
                <Route
                  component={[Function]}
                  path="/properties"
                />
                <Route
                  component={[Function]}
                  path="/keyframes"
                />
              </Switch>
            </div>,
          ],
          "className": "editor-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <NavLink
                  activeClassName="selected"
                  aria-current="page"
                  className="animation-btn"
                  to="/properties"
                >
                  properties
                </NavLink>,
                <NavLink
                  activeClassName="selected"
                  aria-current="page"
                  className="keyframes-btn"
                  to="/keyframes"
                >
                  keyframes
                </NavLink>,
              ],
              "className": "editor-options",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "activeClassName": "selected",
                  "aria-current": "page",
                  "children": "properties",
                  "className": "animation-btn",
                  "to": "/properties",
                },
                "ref": null,
                "rendered": "properties",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "activeClassName": "selected",
                  "aria-current": "page",
                  "children": "keyframes",
                  "className": "keyframes-btn",
                  "to": "/keyframes",
                },
                "ref": null,
                "rendered": "keyframes",
                "type": [Function],
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Switch>
                <Route
                  component={[Function]}
                  path="/properties"
                />
                <Route
                  component={[Function]}
                  path="/keyframes"
                />
              </Switch>,
              "className": "editor",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <Route
                    component={[Function]}
                    path="/properties"
                  />,
                  <Route
                    component={[Function]}
                    path="/keyframes"
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "component": [Function],
                    "path": "/properties",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "component": [Function],
                    "path": "/keyframes",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Link
              className="lower-btn copy-btn"
              replace={false}
              to="/properties/copy"
            >
              copy code
            </Link>,
            <button
              className="lower-btn reset-btn"
              onClick={[Function]}
            >
              reset
            </button>,
            <Link
              className="lower-btn save-as-btn"
              replace={false}
              to="/properties/saveas"
            >
              save as
            </Link>,
            <button
              className="lower-btn save-btn"
              disabled={true}
              onClick={[Function]}
            >
              save
            </button>,
            <Switch>
              <Route
                path="/properties/copy"
                render={[Function]}
              />
              <Route
                path="/properties/saveas"
                render={[Function]}
              />
            </Switch>,
          ],
          "className": "editor-bottom-btns-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "copy code",
              "className": "lower-btn copy-btn",
              "replace": false,
              "to": "/properties/copy",
            },
            "ref": null,
            "rendered": "copy code",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "reset",
              "className": "lower-btn reset-btn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "reset",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "save as",
              "className": "lower-btn save-as-btn",
              "replace": false,
              "to": "/properties/saveas",
            },
            "ref": null,
            "rendered": "save as",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "save",
              "className": "lower-btn save-btn",
              "disabled": true,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "save",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Route
                  path="/properties/copy"
                  render={[Function]}
                />,
                <Route
                  path="/properties/saveas"
                  render={[Function]}
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "path": "/properties/copy",
                  "render": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "path": "/properties/saveas",
                  "render": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="editor-container"
          >
            <div
              className="editor-options"
            >
              <NavLink
                activeClassName="selected"
                aria-current="page"
                className="animation-btn"
                to="/properties"
              >
                properties
              </NavLink>
              <NavLink
                activeClassName="selected"
                aria-current="page"
                className="keyframes-btn"
                to="/keyframes"
              >
                keyframes
              </NavLink>
            </div>
            <div
              className="editor"
            >
              <Switch>
                <Route
                  component={[Function]}
                  path="/properties"
                />
                <Route
                  component={[Function]}
                  path="/keyframes"
                />
              </Switch>
            </div>
          </div>,
          <div
            className="editor-bottom-btns-container"
          >
            <Link
              className="lower-btn copy-btn"
              replace={false}
              to="/properties/copy"
            >
              copy code
            </Link>
            <button
              className="lower-btn reset-btn"
              onClick={[Function]}
            >
              reset
            </button>
            <Link
              className="lower-btn save-as-btn"
              replace={false}
              to="/properties/saveas"
            >
              save as
            </Link>
            <button
              className="lower-btn save-btn"
              disabled={true}
              onClick={[Function]}
            >
              save
            </button>
            <Switch>
              <Route
                path="/properties/copy"
                render={[Function]}
              />
              <Route
                path="/properties/saveas"
                render={[Function]}
              />
            </Switch>
          </div>,
        ],
        "className": "editor-component",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="editor-options"
              >
                <NavLink
                  activeClassName="selected"
                  aria-current="page"
                  className="animation-btn"
                  to="/properties"
                >
                  properties
                </NavLink>
                <NavLink
                  activeClassName="selected"
                  aria-current="page"
                  className="keyframes-btn"
                  to="/keyframes"
                >
                  keyframes
                </NavLink>
              </div>,
              <div
                className="editor"
              >
                <Switch>
                  <Route
                    component={[Function]}
                    path="/properties"
                  />
                  <Route
                    component={[Function]}
                    path="/keyframes"
                  />
                </Switch>
              </div>,
            ],
            "className": "editor-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <NavLink
                    activeClassName="selected"
                    aria-current="page"
                    className="animation-btn"
                    to="/properties"
                  >
                    properties
                  </NavLink>,
                  <NavLink
                    activeClassName="selected"
                    aria-current="page"
                    className="keyframes-btn"
                    to="/keyframes"
                  >
                    keyframes
                  </NavLink>,
                ],
                "className": "editor-options",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "activeClassName": "selected",
                    "aria-current": "page",
                    "children": "properties",
                    "className": "animation-btn",
                    "to": "/properties",
                  },
                  "ref": null,
                  "rendered": "properties",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "activeClassName": "selected",
                    "aria-current": "page",
                    "children": "keyframes",
                    "className": "keyframes-btn",
                    "to": "/keyframes",
                  },
                  "ref": null,
                  "rendered": "keyframes",
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Switch>
                  <Route
                    component={[Function]}
                    path="/properties"
                  />
                  <Route
                    component={[Function]}
                    path="/keyframes"
                  />
                </Switch>,
                "className": "editor",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": Array [
                    <Route
                      component={[Function]}
                      path="/properties"
                    />,
                    <Route
                      component={[Function]}
                      path="/keyframes"
                    />,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "component": [Function],
                      "path": "/properties",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "component": [Function],
                      "path": "/keyframes",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                className="lower-btn copy-btn"
                replace={false}
                to="/properties/copy"
              >
                copy code
              </Link>,
              <button
                className="lower-btn reset-btn"
                onClick={[Function]}
              >
                reset
              </button>,
              <Link
                className="lower-btn save-as-btn"
                replace={false}
                to="/properties/saveas"
              >
                save as
              </Link>,
              <button
                className="lower-btn save-btn"
                disabled={true}
                onClick={[Function]}
              >
                save
              </button>,
              <Switch>
                <Route
                  path="/properties/copy"
                  render={[Function]}
                />
                <Route
                  path="/properties/saveas"
                  render={[Function]}
                />
              </Switch>,
            ],
            "className": "editor-bottom-btns-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "copy code",
                "className": "lower-btn copy-btn",
                "replace": false,
                "to": "/properties/copy",
              },
              "ref": null,
              "rendered": "copy code",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "reset",
                "className": "lower-btn reset-btn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "reset",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "save as",
                "className": "lower-btn save-as-btn",
                "replace": false,
                "to": "/properties/saveas",
              },
              "ref": null,
              "rendered": "save as",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "save",
                "className": "lower-btn save-btn",
                "disabled": true,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "save",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <Route
                    path="/properties/copy"
                    render={[Function]}
                  />,
                  <Route
                    path="/properties/saveas"
                    render={[Function]}
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "path": "/properties/copy",
                    "render": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "path": "/properties/saveas",
                    "render": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
